clippy_lints/src/indexing_slicing.rs

```diff
@@ -141,12 +155,19 @@ impl<'tcx> LateLintPass<'tcx> for IndexingSlicing {
                     (None, None) => return, // [..] is ok.
                 };
 
-                span_lint_and_help(cx, INDEXING_SLICING, expr.span, "slicing may panic", None, help_msg);
+                span_lint_and_then(cx, INDEXING_SLICING, expr.span, "slicing may panic", |diag| {
+                    let note = if cx.tcx.hir().is_inside_const_context(expr.hir_id) {
+                        "the suggestion might not be applicable in constant blocks"
+                    } else {
+                        ""
+                    };
+                    diag.span_suggestion(expr.span, help_msg, note, Applicability::MachineApplicable);
```

Comment:
The applicability is used to indicate if a suggestion can be automatically applied, meaning that the marked span will be replaced by the suggestion. In this case, we don't have an exact suggestion and only a message suggesting the usage of `get(n)` or `get_mut(n)`. Which can't automatically be applied. Therefore, the applicability should be `Unspecified`. This is just some general information.

This is going into the right direction. The `is_inside_const` check is excellent! I was thinking of a different styling, that can be achieved with the following code:

```rs
span_lint_and_then(cx, INDEXING_SLICING, expr.span, "slicing may panic", |diag| {
    diag.help(help_msg);

    if cx.tcx.hir().is_inside_const_context(expr.hir_id) {
        diag.note("the suggestion might not be applicable in constant blocks");
    }
});
```

The same goes for the other `span_lint_and_then`. This one also doesn't require the specification of the applicability. You can also store the note message in a variable, like the current `help_msg` to ensure that both messages stay the same. Then this should be perfect!